File: src/main/java/org/onlineoptimisation/problem/benchmark/EllipsoidalFunction.java
package org.onlineoptimisation.problem.benchmark;

import org.armadillojava.Col;
import org.onlineoptimisation.problem.OptimisationProblem;

public class EllipsoidalFunction extends BenchmarkProblem {

  public EllipsoidalFunction(int numberOfDimensions) {
    super(numberOfDimensions);
  }

  @Override
  public double getObjectiveValueImplementation(Col parameter) {
    double result = 0;
    
    for(int n = 0; n < parameter.n_elem; n++) {
      result += Math.pow(10, 6 * n / (_numberOfDimensions - 1)) * Math.pow(parameter.at(n), 2);
    }
    
    return result;
  }

}

File: src/optimisationProblem/benchmarkOptimisationProblem/blackBoxOptimisationBenchmark/ellipsoidalFunction.cpp
#include "mantella_bits/optimisationProblem/benchmarkOptimisationProblem/blackBoxOptimisationBenchmark/ellipsoidalFunction.hpp"

// C++ standard library
#include <cassert>

// This implementation contains a lot of *magic numbers* and behaviour, introduced by the black-box optimisation benchmark, but only partially explained in the paper.
// So don't expect use to explain the unexplained.
// @see N. Hansen et al., Real-Parameter Black-Box Optimization Benchmarking 2010: Experimental Setup. Research Report RR-7215, INRIA, 2010.
namespace mant {
  namespace bbob {
    EllipsoidalFunction::EllipsoidalFunction(
        const arma::uword numberOfDimensions)
        : BlackBoxOptimisationBenchmark(numberOfDimensions),
          parameterConditioning_(getParameterConditioning(1000000.0)) {
      setParameterTranslation(getRandomParameterTranslation());

      setObjectiveFunction(
          [this](
              const arma::Col<double>& parameter_) {
            assert(parameter_.n_elem == numberOfDimensions_);

            return arma::dot(parameterConditioning_, arma::square(getOscillatedParameter(parameter_)));
          },
          "BBOB Ellipsoidal Function (f2)");
    }
  }
}
